pointer-masking-proposal.adoc

```diff
@@ -295,76 +298,133 @@ MASK
  +
 
 [%header, cols=1*]
 ,===
 spmbase[63:0]
 BASE
 ,===
-Figure 8a: Memory Tagging Extension register (_**spmbase**_) for RV64 (S-MODE) +
+Figure 8c: Memory Tagging Extension register (_**spmbase**_) for RV64 (S-MODE) +
  +
 
 [%header, cols=1*]
 ,===
 upmbase[63:0]
 BASE
 ,===
-Figure 8b: Memory Tagging Extension register (_**upmbase**_) for RV64 (U-MODE) +
+Figure 8d: Memory Tagging Extension register (_**upmbase**_) for RV64 (U-MODE) +
+ +
+
+[%header, cols=1*]
+,===
+mpmbase[127:0]
+BASE
+,===
+Figure 9a: Memory Tagging Extension register (_**mpmbase**_) for RV128 (M-MODE) +
+ +
+
+[%header, cols=1*]
+,===
+hpmbase[127:0]
+BASE
+,===
+Figure 9b: Memory Tagging Extension register (_**hpmbase**_) for RV128 (H-MODE) +
  +
 
 [%header, cols=1*]
 ,===
 spmbase[127:0]
 BASE
 ,===
-Figure 9a: Memory Tagging Extension register (_**spmbase**_) for RV128 (S-MODE) +
+Figure 9c: Memory Tagging Extension register (_**spmbase**_) for RV128 (S-MODE) +
  +
 
 [%header, cols=1*]
 ,===
 upmbase[127:0]
 BASE
 ,===
-Figure 9b: Memory Tagging Extension register (_**upmbase**_) for RV128 (U-MODE) +
+Figure 9d: Memory Tagging Extension register (_**upmbase**_) for RV128 (U-MODE) +
  +
 
 [[explanation]]
 ## Explanation
 
-_**MMTE**_ register fully two-fold function:
+_**MMTE**_ register fulfills two-fold function:
 
-1. Can be programmable by more privileged mode
+1. Can be programmable by more privileged mode (unless PM.Current bit is enabled)
 2. Performs status register function for the current privilege mode
 
-_PM_ bits from **_MMTE_** register are accessible in all modes (U/S/HS/M) and can be read to query the current status of PM feature (if enabled). However, only M-mode privileged code can set the value for _PM_ bits.
+_PM_ bits from **_MMTE_** register are accessible in all modes (U/S/HS/M) and can be read to query the current status of PM feature (if enabled). By default, only higher privileged code can set the value for _PM_ bits. However, higher privileged code can enable _PM.Current_ bit for lower privileged code. In such scenario, current privilege code has a possibility to self-manage its own configuration of _PM_ bits.
+If _PM.Enable_ bit is changing the status, corresponding _xTE_ bit is automatically updated.
+
+_xTE_ bits fulfills two-fold functions:
+
+1. Can be used to quickly disable _PM_ for the corresponding mode without changing current configuration.
+2. Can be used to modify which currently set-up _**xPMMASK**_ / _**xPMBASE**_ and _PM_ bits are enforced.
 
-_UTE_ bit can be read by any mode, but only more privileged modes (S, HS and M) can change the value of it.
+By default, the current CPU mode is using _**xPMMASK**_, _**xPMBASE**_ and _PM_ bits corresponding to it. When CPU is switching the mode, all _xTE_ bits are cleared except the one corresponding to the new mode. _xTE_ bit corresponding to that mode reflects the current configuration of the _PM.Enable_ bit.
```

Comment:
Hypothetical: running in S-mode with PM.Current = 1, PM.Enable = 1, STE = 1, does writing 0 to STE automatically set PM.Enable to 0? In the same scenario does setting STE to 1 automatically sets STE to 1?  

Hypothetical: running in S-mode with PM.Current = 1, PM.Enable = 1, STE = 1, does writing 0 to PM.Enable automatically set STE to 0? In the same scenario does writing PM.Enable to 1 automatically sets STE to 1?  

What happens if I set multiple xTE to 1 at once while PM.Enable = 1?  

What happens if I set all xTE to 0, while PM.Enable = 1?

Reply:
Originally, that was the use-case: bundle xTE with PM.enable. However, after long discussion with Allen this should be updated and do NOT bundle xTE with PM.enable. Getting into account updated docs, scenarios which you have described:

1 and 2 - No. The purpose of xTE bit is to choose which pair of mask/base to use. However, PM.enable bit from the corresponding level defines if they are enforced. During CPU-mode switch appropriate xTE bit is set (and other bits are disabled). However, user can choose to manually enable that xTE bit which he wants. Enforcement of base/mask is done only when PM.enable is set to 1. I think that’s the best way to go

3 - You can have multiple bits to be set. However, only the most significant one has any effect

4 - PM is not enforced in such case. None of the xPMBASE / xPMMASK is being chosen.